### PR TITLE
Add unified search API

### PR DESCRIPTION
**Pull request by ChristophWurst**
See details and design decisions in the code. I felt like it's the better and more discoverable place than putting it into the PR description.

<details>
## Client flow

We decided that the providers should be independent in a way that one slow provider does not slow down the whole search experience, and also that the number of providers register does not influence the total time needed to get all the data.

Therefore the flow for the client (js) code is split into two parts.

1) Get the current list of providers ``
2) Fetch the results for each of the providers concurrently

So we can benefit from the async/concurrent nature of JavaScript in the browser and run multiple search requests in parallel on the server. It's up to the UI code to decide whether the completion of all requests will be awaited or results are shown as they come in.

### Get list of providers

``GET https://localhost/search/providers``

returns

```json
[
  "files",
  "mail"
]
```

### Get individual search results

``GET https://localhost/search/providers/files/search?term=cat``

returns

```json
{
  "name": "Files",
  "isPaginated": false,
  "entries": [
    {
      "thumbnailUrl": "path/to/icon.png",
      "title": "cute cats.jpg",
      "subline": "/Cats",
      "resourceUrl": "/f/21156"
    },
    {
      "thumbnailUrl": "path/to/icon.png",
      "title": "cat 1.jpg",
      "subline": "/Cats",
      "resourceUrl": "/f/21192"
    },
    {
      "thumbnailUrl": "path/to/icon.png",
      "title": "cat 2.jpg",
      "subline": "/Cats",
      "resourceUrl": "/f/25942"
    }
  ],
  "cursor": null
}
```

</details>

Docs are at https://github.com/nextcloud/documentation/pull/2166

---

### CODE REVIEW

File: lib/public/Search/ISearchQuery.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	public function getSortOrder(): int;
+
+	/**
+	 * Get the number of items to return for a paginated result
+	 *
+	 * @return int
+	 * @see \OCP\Search\IProvider for details
+	 * @since 20.0.0
+	 */
+	public function getLimit(): int;
```
Comment (rullzer) [inline lib/public/Search/ISearchQuery.php:69]: Don't you also need an offset then ;)

**Comment by georgehrke** [comment]
A general design question (something that also effects `OCP\Contacts` and `OCP\Calendar`).

Should the user context be provided in the API or are Providers supposed to just use `IUserSession` and take the current user?

File: lib/public/Search/ISearchQuery.php (hunk 1 above)
Comment (ChristophWurst) [reply lib/public/Search/ISearchQuery.php:69]: ``\OCP\Search\ISearchQuery::getCursor`` is your friend :)

**Comment by ChristophWurst** [comment]
> A general design question (something that also effects `OCP\Contacts` and `OCP\Calendar`).
> 
> Should the user context be provided in the API or are Providers supposed to just use `IUserSession` and take the current user?

Very good point. Right now I can not think of a use case where the user is not the one authenticated with the current request but it would make the API a bit more flexible if the user object is provided.

**Comment by rullzer** [comment]
> > A general design question (something that also effects `OCP\Contacts` and `OCP\Calendar`).
> > Should the user context be provided in the API or are Providers supposed to just use `IUserSession` and take the current user?
> 
> Very good point. Right now I can not think of a use case where the user is not the one authenticated with the current request but it would make the API a bit more flexible if the user object is provided.

Yeah I guess providing the user is more clean API wise. :+1:

File: core/Controller/UnifiedSearchController.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+		$this->composer = $composer;
+		$this->userSession = $userSession;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 */
+	public function getProviders(): JSONResponse {
+		return new JSONResponse(
```
Comment (rullzer) [inline core/Controller/UnifiedSearchController.php:58]: I get this from a API point of view. But maybe initial state would be a good idea here?

File: core/routes.php [hunk 3]
```diff
@@ -77,6 +77,8 @@
 		['name' => 'RecommendedApps#index', 'url' => '/core/apps/recommended', 'verb' => 'GET'],
 		['name' => 'Svg#getSvgFromCore', 'url' => '/svg/core/{folder}/{fileName}', 'verb' => 'GET'],
 		['name' => 'Svg#getSvgFromApp', 'url' => '/svg/{app}/{fileName}', 'verb' => 'GET'],
+		['name' => 'UnifiedSearch#getProviders', 'url' => '/search/providers', 'verb' => 'GET'],
+		['name' => 'UnifiedSearch#search', 'url' => '/search/providers/{providerId}/search/{term}', 'verb' => 'GET'],
```
Comment (rullzer) [inline core/routes.php:81]: As discussed

lets move this to a query parameter.

File: core/Controller/UnifiedSearchController.php (hunk 2 above)
Comment (ChristophWurst) [reply core/Controller/UnifiedSearchController.php:58]: Either that and/or we make the providers request cacheable.

**Comment by ChristophWurst** [comment]
- [ ] Blocked by https://github.com/nextcloud/server/pull/21524

**Review by rullzer — APPROVED** [review]

File: apps/comments/lib/Search/LegacyProvider.php [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+use OCP\Search\Provider;
+use function count;
+
+class LegacyProvider extends Provider {
+
+	/**
+	 * Search for $query
+	 *
+	 * @param string $query
+	 * @return array An array of OCP\Search\Result's
```
Comment (kesselb) [inline apps/comments/lib/Search/LegacyProvider.php:43]:
```suggestion
	 * @return array|OCP\Search\Result[] An array of OCP\Search\Result's
```

File: apps/files/lib/Search/FilesSearchProvider.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+use OC\Search\Provider\File;
+use OC\Search\Result\File as FileResult;
+use OCP\IL10N;
+use OCP\IURLGenerator;
+use OCP\IUser;
+use OCP\Search\IProvider;
+use OCP\Search\ISearchQuery;
+use OCP\Search\SearchResult;
+use function array_map as array_mapAlias;
```
Comment (kesselb) [inline apps/files/lib/Search/FilesSearchProvider.php:36]: :thinking:

File: lib/private/Search/SearchQuery.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	/** @var int */
+	private $limit;
+
+	/** @var int|null */
+	private $cursor;
+
+	public function __construct(string $term,
+								int $sortOrder = ISearchQuery::SORT_DATE_DESC,
+								int $limit = self::LIMIT_DEFAULT,
+								?int $cursor = null) {
```
Comment (nickvergessen) [inline lib/private/Search/SearchQuery.php:48]:
```suggestion
								$cursor = null) {
```

Getter docs say `int|string|null`

File: lib/private/Search/SearchQuery.php [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	/** @var string */
+	private $term;
+
+	/** @var int */
+	private $sortOrder;
+
+	/** @var int */
+	private $limit;
+
+	/** @var int|null */
```
Comment (nickvergessen) [inline lib/private/Search/SearchQuery.php:42]:
```suggestion
	/** @var string|int|null */
```

File: core/routes.php [hunk 8]
```diff
@@ -77,6 +77,8 @@
 		['name' => 'RecommendedApps#index', 'url' => '/core/apps/recommended', 'verb' => 'GET'],
 		['name' => 'Svg#getSvgFromCore', 'url' => '/svg/core/{folder}/{fileName}', 'verb' => 'GET'],
 		['name' => 'Svg#getSvgFromApp', 'url' => '/svg/{app}/{fileName}', 'verb' => 'GET'],
+		['name' => 'UnifiedSearch#getProviders', 'url' => '/search/providers', 'verb' => 'GET'],
```
Comment (nickvergessen) [inline core/routes.php:80]: Can these be OCS so it's easier to use for clients and API consumers?

**Review by kesselb — APPROVED** [review]

File: apps/comments/lib/Search/LegacyProvider.php (hunk 4 above)
Comment (ChristophWurst) [reply apps/comments/lib/Search/LegacyProvider.php:43]: that is the old code. I only moved it. soooooo don't care :speak_no_evil:

File: core/routes.php (hunk 8 above)
Comment (nickvergessen) [reply core/routes.php:80]: So we decided to delay OCS, until the frontend is there, so we are more sure it "works" and can do everything we need.

**Review by nickvergessen — APPROVED** [review]
🐘

**Comment by ChristophWurst** [comment]
`@since` annotations were missing

**Comment by ChristophWurst** [comment]
And now a few tests failed :roll_eyes:

**Comment by skjnldsv** [comment]
I found an issue @ChristophWurst 
- [ ] Files limit is [not taken into account](https://github.com/nextcloud/server/blob/27b8637e3e2676e81df74891043c39530bdc1cda/lib/private/Search/Provider/File.php#L46)

**Comment by ChristophWurst** [comment]
> Files limit is [not taken into account](https://github.com/nextcloud/server/blob/27b8637e3e2676e81df74891043c39530bdc1cda/lib/private/Search/Provider/File.php#L46)

This is about the current files search implementation, right? It's not a paginated one, so there is no limit per page. A limit only makes sense if you break down the result set into chunks as pages. When the provider does not support that, the unification logic is out of luck. You need to factor that in on the UI end.

Summary
* Full results have an arbitrary number of entries
* Paginated results should adhere to the specified limit, thus only return a certain number of entries

> You're using OCP\Search\Provider & OC\Search\Result\File that are deprecated

*This PR* deprecated them :see_no_evil: The files search provider is something that has to be redone. I just used the existing code and plugged it into the new one. This needs storage experts knowledge to transform properly, I lack that.

**Comment by skjnldsv** [comment]
> It's not a paginated one, so there is no limit per page. A limit only makes sense if you break down the result set into chunks as pages

And I can't enforce that?
Part of the spec is being able to paginate every results. I need to be able to show only 5 per types

**Comment by ChristophWurst** [comment]
As I tried to explain, no, you can't.

The only thing we could do for this is to cap the results at a certain point for those providers that are not paginated. But then the user will never see the full list.

> Part of the spec is being able to paginate every results.

But you can still do this. Just because the server gives you more data does not mean you have to show everything? Why not store the results as you get them, show the first x entries and when you would have to get the next page for a paginated provider you just show more of the already known entries.

It's a bit of additional work on the front-end, sure, but that is the best we can do if we want to support paginated and non-paginated providers.

**Comment by ChristophWurst** [comment]
Docs at https://github.com/nextcloud/documentation/pull/2166

**Pull request merged**